### PR TITLE
tap: implement streaming tap for the transport socket

### LINKED ISSUE #6127: IpVersions/SslTapIntegrationTest tests flaky
Looks like we had about 7 CI failures overnight, and when I spot checked half they were all the ssl_integration_test

Two example failures from 2 different runs:

https://circleci.com/:vcs_type/organizations/:org/settings[ RUN      ] IpVersions/SslTapIntegrationTest.RequestWithJsonBodyAsString/IPv4
unknown file: Failure
C++ exception with description "Unable to parse JSON as proto (INVALID_ARGUMENT:Unexpected end of string. Expected a value.

^): " thrown in the test body.
[  FAILED  ] IpVersions/SslTapIntegrationTest.RequestWithJsonBodyAsString/IPv4, where GetParam() = 4-byte object <00-00 00-00> (298 ms)

[ RUN      ] IpVersions/SslTapIntegrationTest.TruncationWithMultipleDataFrames/IPv4
test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:505: Failure
Expected equality of these values:
  trace.socket_buffered_trace().events().size()
    Which is: 0
  2
[  FAILED  ] IpVersions/SslTapIntegrationTest.TruncationWithMultipleDataFrames/IPv4, where GetParam() = 4-byte object <00-00 00-00> (109 ms)

### PR DESCRIPTION
**Pull request by mattklein123**
Risk Level: Medium for existing tap transport users, none for anyone else
Testing: New UTs
Docs Changes: Added
Release Notes: N/A
Fixes #6127

### CODE REVIEW

File: test/extensions/transport_sockets/tap/tap_config_impl_test.cc [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    EXPECT_CALL(*config_, streaming()).WillRepeatedly(Return(streaming));
+    EXPECT_CALL(*config_, maxBufferedRxBytes()).WillRepeatedly(Return(1024));
+    EXPECT_CALL(*config_, maxBufferedTxBytes()).WillRepeatedly(Return(1024));
+    EXPECT_CALL(*config_, timeSource()).WillRepeatedly(ReturnRef(time_system_));
+    time_system_.setSystemTime(std::chrono::seconds(0));
+    tapper_ = std::make_unique<PerSocketTapperImpl>(config_, connection_);
+  }
+
+  std::shared_ptr<MockSocketTapConfig> config_{std::make_shared<MockSocketTapConfig>()};
+  TapCommon::MockPerTapSinkHandleManager* sink_manager_ =
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/extensions/transport_sockets/tap/tap_config_impl_test.cc]: This is returned as a unique_ptr so it needs to be a raw pointer. I will add a comment.

File: test/extensions/common/tap/common.h [hunk 2]
```diff
@@ -20,6 +20,13 @@ std::ostream& operator<<(std::ostream& os, const TraceWrapper& trace) {
 
 namespace Envoy {
 namespace Extensions {
+
+MATCHER_P(TraceEqual, rhs, "") {
+  envoy::data::tap::v2alpha::TraceWrapper expected_trace;
+  MessageUtil::loadFromYaml(rhs, expected_trace);
+  return arg.DebugString() == expected_trace.DebugString();
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/extensions/common/tap/common.h]: I'm going to punt on making this generic for right now, but will fix the message differencer part. The issue with making it generic is the target message for the YAML needs to be templatized. I think this is doable but from 5 minutes is trying it's not obvious so I will add a TODO and look at it later.

**Comment by mattklein123** [comment]
@lizan @htuch updated

**Comment by mattklein123** [comment]
@htuch I'm going to merge to fix the flake. If you have any comments I will fix them in a follow up.

**Comment by htuch** [comment]
This look fine.

**Pull request merged**